src/Npgsql.DependencyInjection/NpgsqlServiceCollectionExtensions.cs

```diff
@@ -0,0 +1,101 @@
+    /// <param name="dataSourceBuilderAction">
+    /// An action to configure the <see cref="NpgsqlDataSourceBuilder" /> for further customizations of the <see cref="NpgsqlDataSource" />.
+    /// </param>
+    /// <param name="connectionLifetime">
+    /// The lifetime with which to register the <see cref="NpgsqlConnection" /> in the container.
+    /// Defaults to <see cref="ServiceLifetime.Scoped" />.
+    /// </param>
+    /// <param name="dataSourceLifetime">
+    /// The lifetime with which to register the <see cref="NpgsqlDataSource" /> service in the container.
+    /// Defaults to <see cref="ServiceLifetime.Singleton" />.
+    /// </param>
+    /// <returns>The same service collection so that multiple calls can be chained.</returns>
+    public static IServiceCollection AddNpgsqlDataSource(
+        this IServiceCollection serviceCollection,
+        string connectionString,
+        Action<NpgsqlDataSourceBuilder> dataSourceBuilderAction,
+        ServiceLifetime connectionLifetime = ServiceLifetime.Scoped,
+        ServiceLifetime dataSourceLifetime = ServiceLifetime.Singleton)
+        => AddNpgsqlDataSourceCore(serviceCollection, connectionString, dataSourceBuilderAction, connectionLifetime, dataSourceLifetime);
+
+    /// <summary>
+    /// Registers an <see cref="NpgsqlDataSource" /> and an <see cref="NpgsqlConnection" /> in the <see cref="IServiceCollection" />,
+    /// </summary>
+    /// <param name="serviceCollection">The <see cref="IServiceCollection" /> to add services to.</param>
+    /// <param name="connectionString">An Npgsql connection string.</param>
+    /// <param name="connectionLifetime">
+    /// The lifetime with which to register the <see cref="NpgsqlConnection" /> in the container.
+    /// Defaults to <see cref="ServiceLifetime.Scoped" />.
+    /// </param>
+    /// <param name="dataSourceLifetime">
+    /// The lifetime with which to register the <see cref="NpgsqlDataSource" /> service in the container.
+    /// Defaults to <see cref="ServiceLifetime.Singleton" />.
+    /// </param>
+    /// <returns>The same service collection so that multiple calls can be chained.</returns>
+    public static IServiceCollection AddNpgsqlDataSource(
+        this IServiceCollection serviceCollection,
+        string connectionString,
+        ServiceLifetime connectionLifetime = ServiceLifetime.Scoped,
+        ServiceLifetime dataSourceLifetime = ServiceLifetime.Singleton)
+        => AddNpgsqlDataSourceCore(
+            serviceCollection, connectionString, dataSourceBuilderAction: null, connectionLifetime, dataSourceLifetime);
+
+    static IServiceCollection AddNpgsqlDataSourceCore(
+        this IServiceCollection serviceCollection,
+        string connectionString,
+        Action<NpgsqlDataSourceBuilder>? dataSourceBuilderAction,
+        ServiceLifetime connectionLifetime = ServiceLifetime.Scoped,
+        ServiceLifetime dataSourceLifetime = ServiceLifetime.Singleton)
+    {
+        serviceCollection.Add(
```

Comment:
It's common/best practice to use the TryAdd versions to allow users to add their own versions of some of the dependencies before they would call `AddX` without overriding them (I believe last registration wins when it comes to the default pick by sp.GetService). At the very least - ideally they would all be TryAdd - we should be doing this for the base classes as users could already have another DbConnection/DbDataSource derived type registered.

Reply:
Good point. It sounds quite weird for someone to have both a non-Npgsql DbDataSource and a NpgsqlDataSource registered, but I agree it makes sense to make them all TryAdd.